File: src/main/java/pl/training/module7/parameterobject/after/RolloutPlanner.java
package pl.training.module7.parameterobject.after;

import java.util.Objects;

public final class RolloutPlanner {
    private static final long DEPLOYMENT_SECONDS_PER_BATCH = 30;

    public long estimateSeconds(RolloutSpec spec) {
        Objects.requireNonNull(spec, "spec");

        long batches = Math.ceilDiv(
                (long) spec.instances(),
                spec.batchSize());
        long deploymentSeconds = Math.multiplyExact(
                batches,
                DEPLOYMENT_SECONDS_PER_BATCH);
        long pauseTime = Math.multiplyExact(
                batches - 1,
                (long) spec.pauseSeconds());
        return Math.addExact(deploymentSeconds, pauseTime);
    }

    public String describe(RolloutSpec spec) {
        Objects.requireNonNull(spec, "spec");

        return "service=" + spec.service()
                + ";region=" + spec.region()
                + ";instances=" + spec.instances()
                + ";batchSize=" + spec.batchSize()
                + ";pauseSeconds=" + spec.pauseSeconds();
    }
}

File: typescript/src/module7/parameterobject/after/RolloutPlanner.ts
import { requireNonNull } from '../../../shared/requireNonNull.js';
import { addExact, ceilDiv, multiplyExact } from '../../ExactMath.js';
import type { RolloutSpec } from './RolloutSpec.js';

export class RolloutPlanner {
  private static readonly DEPLOYMENT_SECONDS_PER_BATCH = 30;

  estimateSeconds(spec: RolloutSpec): number {
    requireNonNull(spec, 'spec');

    const batches = ceilDiv(
      spec.instances,
      spec.batchSize);
    const deploymentSeconds = multiplyExact(
      batches,
      RolloutPlanner.DEPLOYMENT_SECONDS_PER_BATCH);
    const pauseTime = multiplyExact(
      batches - 1,
      spec.pauseSeconds);
    return addExact(deploymentSeconds, pauseTime);
  }

  describe(spec: RolloutSpec): string {
    requireNonNull(spec, 'spec');

    return 'service=' + spec.service
      + ';region=' + spec.region
      + ';instances=' + spec.instances
      + ';batchSize=' + spec.batchSize
      + ';pauseSeconds=' + spec.pauseSeconds;
  }
}
